File: Assets/AboutXLua/Scripts/Core/LogSystem/LogUtility.cs
using System;
using System.Collections;
using System.Collections.Generic;

public static class LogUtility
{
    private static List<LogEntry> _logEntries = new List<LogEntry>();
    private static readonly object _lockObj = new object();
    
    // 日志级别开关（可在运行时动态调整）
    public static bool EnableInfoLogs = true;
    public static bool EnableWarningLogs = true;
    public static bool EnableErrorLogs = true;
    
    public static IReadOnlyList<LogEntry> LogEntries 
    { 
        get 
        { 
            lock (_lockObj)
            {
                return new List<LogEntry>(_logEntries);
            }
        } 
    }
    
    public static event Action OnLogAdded;
    
    public static void ClearLogs()
    {
        lock (_lockObj)
        {
            _logEntries.Clear();
        }
    }
    
    public static void Log(LogLayer layer, string source, LogLevel level, string message)
    {
        // 根据级别开关决定是否记录
        if ((level == LogLevel.Info && !EnableInfoLogs) ||
            (level == LogLevel.Warning && !EnableWarningLogs) ||
            (level == LogLevel.Error && !EnableErrorLogs))
        {
            return;
        }
        
        string formatted = $"[Layer:{layer}][Script:{source}][{level}] {message}";

        // 记录日志到列表
        lock (_lockObj)
        {
            _logEntries.Add(new LogEntry
            {
                ScriptType = "CSharp",
                Layer = layer,
                Level = level,
                Source = source,
                Message = message,
                FormattedMessage = formatted,
                Time = DateTime.Now
            });
        }
        OnLogAdded?.Invoke();
        
        switch (level)
        {
            case LogLevel.Info:
                UnityEngine.Debug.Log(formatted);
                break;
            case LogLevel.Warning:
                UnityEngine.Debug.LogWarning(formatted);
                break;
            case LogLevel.Error:
                UnityEngine.Debug.LogError(formatted);
                break;
        }
    }
    
    // 快捷方法 - Info
    public static void Info(LogLayer layer, string source, string message)
    {
        Log(layer, source, LogLevel.Info, message);
    }
    
    // 快捷方法 - Warning
    public static void Warning(LogLayer layer, string source, string message)
    {
        Log(layer, source, LogLevel.Warning, message);
    }
    
    // 快捷方法 - Error
    public static void Error(LogLayer layer, string source, string message)
    {
        Log(layer, source, LogLevel.Error, message);
    }
    
    // 用于 Lua 调用（需绑定到 LuaCallCSharp）
    public static void LogFromLua(int layer, string source, int level, string message)
    {
        // 根据级别开关决定是否记录
        if ((level == (int)LogLevel.Info && !EnableInfoLogs) ||
            (level == (int)LogLevel.Warning && !EnableWarningLogs) ||
            (level == (int)LogLevel.Error && !EnableErrorLogs))
        {
            return;
        }
        
        if (!Enum.IsDefined(typeof(LogLayer), layer)) layer = (int)LogLayer.Custom;
        if (!Enum.IsDefined(typeof(LogLevel), level)) level = (int)LogLevel.Info;
        
        var logLayer = (LogLayer)layer;
        var logLevel = (LogLevel)level;
        var fullSource = "Lua:" + source;
        string formatted = $"[Layer:{logLayer}][Script:{fullSource}][{logLevel}] {message}";

        // 记录Lua日志到列表
        lock (_lockObj)
        {
            _logEntries.Add(new LogEntry
            {
                ScriptType = "Lua",
                Layer = logLayer,
                Level = logLevel,
                Source = fullSource,
                Message = message,
                FormattedMessage = formatted,
                Time = DateTime.Now
            });
        }
        OnLogAdded?.Invoke();

        switch (logLevel)
        {
            case LogLevel.Info:
                UnityEngine.Debug.Log(formatted);
                break;
            case LogLevel.Warning:
                UnityEngine.Debug.LogWarning(formatted);
                break;
            case LogLevel.Error:
                UnityEngine.Debug.LogError(formatted);
                break;
        }
    }
}

public enum LogLevel
{
    Error,
    Warning,
    Info
}

public enum LogLayer {
    Core,       // 底层核心
    Framework,  // 中间层
    Game,         // 上层展示
    Global,      // 全局
    Custom      // 可自定义扩展
}

public class LogEntry
{
    public string ScriptType { get; set; } // Lua或CSharp
    public LogLayer Layer { get; set; }
    public LogLevel Level { get; set; }
    public string Source { get; set; }
    public string Message { get; set; }
    public string FormattedMessage { get; set; }
    public DateTime Time { get; set; }
}

File: Assets/AboutXLua/LuaScripts/Core/LogSystem/LogUtility.lua
-- 日志工具
local LogUtility = {}

-- C#中定义顺序就是这个
LogUtility.LogLevel = {
    Error = 0,
    Warning = 1,
    Info = 2,    
}

LogUtility.LogLayer = {
    Core = 0,
    Framework = 1,
    Game = 2,
    Global = 3,
    Custom = 4,
}

-- 格式化打印
local function format_message(layer, source, level, msg)
    local levelStr = ({ "Info", "Warning", "Error" })[level + 1] or "Unknown"
    local layerStr = ({ "Core", "Framework", "Game","Global","Custom" })[layer + 1] or "Unknown"
    return string.format("[Layer:%s][LuaScript:%s][%s] %s", layerStr, source, levelStr, msg)
end

-- 打印核心方法（优先使用C#日志，print作为备选）
function LogUtility.Print(layer, source, level, msg)
    local formatted = format_message(layer, source, level, msg)

    -- 优先调用C#日志接口
    if CS and CS.LogUtility then
        CS.LogUtility.LogFromLua(layer, source, level, msg)
    else
        -- print作为备选方案
        if level == LogUtility.LogLevel.Info then
            print(formatted)
        elseif level == LogUtility.LogLevel.Warning then
            print("<color=yellow>" .. formatted .. "</color>")
        elseif level == LogUtility.LogLevel.Error then
            print("<color=red>" .. formatted .. "</color>")
        end
    end
end

-- 快捷方法
function LogUtility.Info(layer, source, msg)
    LogUtility.Print(layer, source, LogUtility.LogLevel.Info, msg)
end

function LogUtility.Warning(layer, source, msg)
    LogUtility.Print(layer, source, LogUtility.LogLevel.Warning, msg)
end

function LogUtility.Error(layer, source, msg)
    LogUtility.Print(layer, source, LogUtility.LogLevel.Error, msg)
end

return LogUtility